File: src/movegen/occupancy.rs
use crate::bitboard::clear_bit;

pub fn mask_occupancy(index: u64, bits_in_mask: u8, attack_mask: u64) -> u64 {
    let mut attack_mask = attack_mask;

    let mut occupancy = 0u64;

    for count in 0..bits_in_mask {
        let square = attack_mask.trailing_zeros() as u8;

        clear_bit(&mut attack_mask, square);

        if (index & (1u64 << count)) != 0 {
            occupancy |= 1u64 << square;
        }
    }

    occupancy
}

File: internal/movegen/occupancy.go
package movegen

import (
	"github.com/notmalte/psce/internal/bitboard"
	"math/bits"
)

func MaskOccupancy(attackMask uint64, index uint64) uint64 {
	occupancy := uint64(0)
	bitsInMask := bits.OnesCount64(attackMask)

	for count := 0; count < bitsInMask; count++ {
		square := uint8(bits.TrailingZeros64(attackMask))
		bitboard.ClearBit(&attackMask, square)

		if bitboard.GetBit(index, uint8(count)) {
			bitboard.SetBit(&occupancy, square)
		}
	}

	return occupancy
}
